File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-3.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    private fun serialize(root: TreeNode?): String {
        return when (root) {
            null -> "$#"
            else -> "$${root.`val`}${serialize(root.left)}${serialize(root.right)}"
        }
    }

    private fun zFunction(s: String): IntArray {
        val n = s.length
        val z = IntArray(n)
        var l = 0
        var r = 0

        for (i in 1 until n) {
            if (i <= r) {
                z[i] = minOf(r - i + 1, z[i - l])
            }
            while (i + z[i] < n && s[z[i]] == s[i + z[i]]) {
                z[i]++
            }
            if (i + z[i] - 1 > r) {
                l = i
                r = i + z[i] - 1
            }
        }
        return z
    }

    fun isSubtree(root: TreeNode?, subRoot: TreeNode?): Boolean {
        val serializedRoot = serialize(root)
        val serializedSubRoot = serialize(subRoot)
        val combined = serializedSubRoot + "|" + serializedRoot

        val zValues = zFunction(combined)
        val subLen = serializedSubRoot.length

        return (subLen + 1 until combined.length).any { i -> zValues[i] == subLen }
    }
}

File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-3.cpp
// DFS
class Solution {
public:
    vector<vector<int>> result;

    vector<vector<int>> levelOrder(TreeNode* root) {
        dfs(root, 0);
        return result;
    }

    void dfs(TreeNode* node, int depth) {
        if(node == nullptr) return;

        if(result.size() == depth) {
            result.push_back(vector<int>());
        }
        result[depth].push_back(node->val);
        dfs(node->left, depth + 1);
        dfs(node->right, depth + 1);
    }
};
